parlai/tasks/dailydialog/agents.py

```diff
@@ -0,0 +1,91 @@
+#!/usr/bin/env python3
+
+# Copyright (c) 2017-present, Facebook, Inc.
+# All rights reserved.
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree. An additional grant
+# of patent rights can be found in the PATENTS file in the same directory.
+
+"""
+Daily Dialog
+https://arxiv.org/abs/1710.03957
+
+Original data is copyright by the owners of the paper, and free for use in research.
+
+Every conversation contains entries with special fields (see the paper):
+
+- emotion
+- act_type
+- topic
+
+This teacher plays both sides of the conversation.
+"""
+
+import os
+import json
+from parlai.core.teachers import FixedDialogTeacher
+from .build import build
+
+
+START_ENTRY = {
+    'text': '__SILENCE__',
+    'emotion': 'no_emotion',
+    'act': 'no_act',
+}
+
+
+class DefaultTeacher(FixedDialogTeacher):
+    def __init__(self, opt, shared=None):
+        super().__init__(opt, shared)
+        self.opt = opt
+        if shared:
+            self.data = shared['data']
+        else:
+            build(opt)
+            fold = opt.get('datatype', 'train').split(':')[0]
+            self._setup_data(fold)
+
+        self.num_exs = sum(len(d['dialogue']) for d in self.data)
+
+        # we play both sides
+        self.num_eps = 2 * len(self.data)
+        self.reset()
+
+    def num_episodes(self):
+        return self.num_eps
+
+    def num_examples(self):
+        return self.num_exs
+
+    def _setup_data(self, fold):
+        self.data = []
+        fpath = os.path.join(self.opt['datapath'], 'dailydialog', fold + '.json')
+        with open(fpath) as f:
+            for line in f:
+                self.data.append(json.loads(line))
+
+    def get(self, episode_idx, entry_idx=0):
+        # we play both sides!
```

Comment:
can you elaborate on how this setup enforces playing both sides?

Reply:
So the conversation doesn't have a canonical Speaker 1 and Speaker 2, and the evaluation in the original paper seems to generally evaluate responses. So in this Teacher, I've made it that on even episode_idx, we play Speaker 1, and on odd episode_idx, we play Speaker 2, so you learn from both sides of the conversation.

Reply:
Instead of doing this, shouldn't we effectively double the data by having each episode appear twice (once where we are Speaker1, and once where we are Speaker2)? This is what I did for ConvAI2

Reply:
I do, I just don't store the data twice.

Reply:
I see...is the number of episodes odd? Otherwise the speaker ID will remain the same for each episode

Reply:
No, b/c:

https://github.com/facebookresearch/ParlAI/blob/ff72628bc6f3832704b4f2a74660846e94243e42/parlai/tasks/dailydialog/agents.py#L51

So I "double" the amount of episodes in the data, then alternate between which speaker I play

Reply:
Ahh ok I think I see, it's the `episode_idx // 2` that makes it so you see each example twice right?